drivers/st25r3911b/st25r3911b.c

```diff
@@ -0,0 +1,552 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <kernel.h>
+#include <misc/byteorder.h>
+#include <device.h>
+#include <soc.h>
+#include <gpio.h>
+#include <logging/log.h>
+#include <drivers/st25r3911b_spi.h>
+#include <drivers/st25r3911b_interrupt.h>
+#include <drivers/st25r3911b.h>
+
+LOG_MODULE_DECLARE(st25r3911b);
+
+#define T_OSC_STABLE 10
+#define T_POWER_SUPP_MEAS 100
+#define T_CA_TIMEOUT 10
+#define T_COMMON_CMD 10
+
+#define POWER_SUPP_3V3_MAX_LVL 3600
+#define POWER_SUPP_3V3_INIT_VAL 2400
+#define POWER_SUPP_3V3_STEP 100
+#define POWER_SUPP_5V_INIT_VAL 3900
+#define POWER_SUPP_5V_STEP 120
+#define POWER_SUPP_MEAS_INTEGER 23
+#define POWER_SUPP_MEAS_FRACTION 438
+#define POWER_SUPP_MEAS_FRACTION_DIVISOR 100
+#define REGULATOR_INIT_REG_VALUE 5
+
+#define FIFO_TX_WATER_LVL_32 32
+#define FIFO_TX_WATER_LVL_16 16
+#define FIFO_RX_WATER_LVL_80 80
+#define FIFO_RX_WATER_LVL_64 64
+
+#define FIFO_TX_WATER_16_EMPTY (ST25R3911B_MAX_FIFO_LEN - FIFO_TX_WATER_LVL_16)
+#define FIFO_TX_WATER_32_EMPTY (ST25R3911B_MAX_FIFO_LEN - FIFO_TX_WATER_LVL_32)
+
+static struct device *gpio_dev;
+
+static int osc_start(void)
+{
+	int err;
+	u8_t reg = 0;
+	u32_t status;
+
+	err = st25r3911b_reg_read(ST25R3911B_REG_OP_CTRL, &reg);
+	if (err) {
+		return err;
+	}
+
+	if (!(reg & ST25R3911B_REG_OP_CTRL_EN)) {
+
+		LOG_DBG("Oscilator is disabled, enabling it");
+
+		err = st25r3911b_irq_enable(ST25R3911B_IRQ_MASK_OSC);
+		if (err) {
+			return err;
+		}
+
+		err = st25r3911b_reg_modify(ST25R3911B_REG_OP_CTRL,
+					    0, ST25R3911B_REG_OP_CTRL_EN);
+		if (err) {
+			return err;
```

Comment:
in case of error disable irq that was enabled by `st25r3911b_irq_enable`

Reply:
similarly in other places

Reply:
Fixed